### DISCUSSION TITLE
Using both MC and FW servos on targets where timers support it

### DISCUSSION THREAD

**Discussion by eksperts** [iNavFlight/inav#7956] [General]
Note: all of research below has been done on stable 4.1.

Objective: possibility to use the same FC for plane and multirotor platforms, with the spare/unused motor outputs assigned to drive servos.

Here's a real-world example. I have an FC that can do 8 motor outputs, but I only need 4 for a Quad X mixer. However, I also need to drive 3 additional servos on the quad, which means that IF the timers permit it, I should be able to do so.
However, the current target configuration explicitly states:
a) 8 Quad motors. OR--
b) 4 Wing motors + 4 Wing servos.
Which means that if I set up a Quad mixer, none of the motor outputs can be used as servos, except perhaps with SERVO_TILT which only supports up to 2 servos.
This is confusing, because if a Wing can do 4 motors and 4 servos, then a Quad should be able to do the same?

Target: TMOTORF7V2
Board: T-Motor F7 Pro
The current `target.c` file says:
```
DEF_TIM(TIM3, CH3, PB0, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S1
DEF_TIM(TIM3, CH4, PB1, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S2
DEF_TIM(TIM3, CH1, PB4, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S3
DEF_TIM(TIM3, CH2, PB5, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S4 
DEF_TIM(TIM4, CH1, PB6, TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO, 0, 0),  // S5
DEF_TIM(TIM4, CH2, PB7, TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO, 0, 0),  // S6
DEF_TIM(TIM8, CH3, PC8, TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO, 0, 0),  // S7
DEF_TIM(TIM8, CH4, PC9, TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO, 0, 0),  // S8 
```

I have now changed the S5 - S8 pins to use `MC_SERVO` instead of `FW_SERVO`:
```
DEF_TIM(TIM3, CH3, PB0, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S1
DEF_TIM(TIM3, CH4, PB1, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S2
DEF_TIM(TIM3, CH1, PB4, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S3
DEF_TIM(TIM3, CH2, PB5, TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR, 0, 0),  // S4 
DEF_TIM(TIM4, CH1, PB6, TIM_USE_MC_MOTOR | TIM_USE_MC_SERVO, 0, 0),  // S5
DEF_TIM(TIM4, CH2, PB7, TIM_USE_MC_MOTOR | TIM_USE_MC_SERVO, 0, 0),  // S6
DEF_TIM(TIM8, CH3, PC8, TIM_USE_MC_MOTOR | TIM_USE_MC_SERVO, 0, 0),  // S7
DEF_TIM(TIM8, CH4, PC9, TIM_USE_MC_MOTOR | TIM_USE_MC_SERVO, 0, 0),  // S8 
```

And as a result, my custom build does, indeed, work as a Quad X with 3 (maybe even 4 if I tried) servos.
However, such change to the target means that this build would not be suitable for planes as the outputs can no longer be assigned to `FW` mixer.
To add to this, it should still be possible to use the same FC for a Hexa with 2 servos, or a full X8/octa without any servos at all. So there's a lot of potentially perfectly valid configurations here.

I have been told that allowing both `MC_SERVO` and `FW_SERVO` on the same line would pose a risk, so I'm not suggesting anything like this:
```
DEF_TIM(TIM4, CH1, PB6, TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO | TIM_USE_MC_SERVO, 0, 0),  // S5
```
But someone with experience could maybe confirm if that would work and still be safe.

Ideally, I'd like to see an option to use the free/spare motor outputs as servo outputs on both Plane and Quad configs in an either-or manner: so that I can have `FW_SERVO`s on Wing mixer and `MC_SERVO`s on Quad.
As I understand it, such capability directly depends on the timers a given FC is running on the motor output pins. Meaning, if we are able to see that there is no conflict and we won't be running both motors and servos on the same timers, it should be fine.

I have tried to use a Custom mixer, but then then aircraft seems to not understand what it is and what it should do to fly even with the right Mixer settings, so I had to abandon that route.
And of course, this likely affects all of the targets that have a compatible timer configuration (not just TMOTORF7V2), which then might mean changes in both the `target.c` files and the UI of the Mixer tab to accommodate the various configuration options.

Any further input welcome.

**Comment by zvikaf** [comment]
Excellent question... did you find a solution ?

**Reply by eksperts** [reply]
No, I did not

**Reply by zvikaf** [reply]
Bummer :-(
